libraries/chain/controller.cpp

```diff
@@ -1257,6 +1246,57 @@ struct controller_impl {
       }
    }
 
+   void transition_to_savanna() {
+      assert(chain_head.header().contains_header_extension(instant_finality_extension::extension_id()));
+      // copy head branch branch from legacy forkdb legacy to savanna forkdb
+      fork_database_legacy_t::branch_t legacy_branch;
+      block_state_legacy_ptr legacy_root;
+      fork_db.apply_l<void>([&](const auto& forkdb) {
+         legacy_root = forkdb.root();
+         legacy_branch = forkdb.fetch_branch(forkdb.head()->id());
+      });
+
+      assert(!!legacy_root);
+      assert(read_mode == db_read_mode::IRREVERSIBLE || !legacy_branch.empty());
+      ilog("Transitioning to savanna, IF Genesis Block ${gb}, IF Critical Block ${cb}", ("gb", legacy_root->block_num())("cb", chain_head.block_num()));
+      auto new_root = block_state::create_if_genesis_block(*legacy_root);
+      fork_db.switch_from_legacy(new_root);
+      fork_db.apply_s<void>([&](auto& forkdb) {
+         block_state_ptr prev = forkdb.root();
+         for (auto bitr = legacy_branch.rbegin(); bitr != legacy_branch.rend(); ++bitr) {
+            const bool skip_validate_signee = true; // validated already
+            auto new_bsp = std::make_shared<block_state>(
+                  *prev,
+                  (*bitr)->block,
+                  protocol_features.get_protocol_feature_set(),
+                  validator_t{}, skip_validate_signee);
+            // legacy_branch is from head, all should be validated
+            assert((*bitr)->action_receipt_digests);
+            auto action_mroot = calculate_merkle(*(*bitr)->action_receipt_digests);
+            // Create the valid structure for producing
+            new_bsp->valid = prev->new_valid(*new_bsp, action_mroot);
+            forkdb.add(new_bsp, (*bitr)->is_valid() ? mark_valid_t::yes : mark_valid_t::no, ignore_duplicate_t::yes);
```

Comment:
Yes, except when running in irreversible mode. This is needed for support for irreversible mode. See #2335

Reply:
But we can still do `skip_validate_signee = true;` on line 1267 above?

Reply:
Yes, because we will validate when apply() is called. apply() is called for it to be in head if non-irreversible, or will be called as it is being made LIB in irreversible mode.

Reply:
I must be missing something. Does the validate happen between these lines? 
![image](https://github.com/AntelopeIO/leap/assets/1496907/81a1b907-eeaa-442f-a517-c28db19438bb)

Reply:
No, sorry, I'm not being clear.

Might be easier to understand in the #2335 
https://github.com/AntelopeIO/leap/pull/2335/files#diff-42e9f97eed543dd784b1b77a536c50ff8e2403d5bbd3b1fb7f5dd90201391061L1273-L1278

The signature is always validated by `apply()`, the blocks are not validated when put into the forkdb when running in irreversible mode. They are already validated and part of the head branch when running in non-irreversible mode.